src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -0,0 +1,364 @@
+                {
+                    CERT_CONTEXT* pCertContext = hCertContext.DangerousGetCertContext();
+                    CERT_INFO* pCertInfo = pCertContext->pCertInfo;
+
+                    CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO* pEncodeInfo = (CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO*)(hb.Alloc(sizeof(CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO)));
+
+                    pEncodeInfo->cbSize = sizeof(CMSG_KEY_TRANS_RECIPIENT_ENCODE_INFO);
+
+                    CRYPT_ALGORITHM_IDENTIFIER algId = pCertInfo->SubjectPublicKeyInfo.Algorithm;
+                    pEncodeInfo->KeyEncryptionAlgorithm = algId;
+
+                    pEncodeInfo->pvKeyEncryptionAuxInfo = IntPtr.Zero;
+                    pEncodeInfo->hCryptProv = IntPtr.Zero;
+
+                    pEncodeInfo->RecipientPublicKey = pCertInfo->SubjectPublicKeyInfo.PublicKey;
+
+                    pEncodeInfo->RecipientId = EncodeRecipientId(recipient, hCertContext, pCertContext, pCertInfo, hb);
+
+                    return pEncodeInfo;
+                }
+            }
+
+            //
+            // This returns an allocated native memory block. Its lifetime (and that of any allocated subblocks it may point to) is that of "hb". 
+            //
+            private static unsafe CMSG_KEY_AGREE_RECIPIENT_ENCODE_INFO* EncodeKeyAgreeRecipientInfo(CmsRecipient recipient, AlgorithmIdentifier contentEncryptionAlgorithm, HeapBlockRetainer hb)
+            {
+                // "recipient" is a deep-cloned CmsRecipient object whose lifetime this class controls. Because of this, we can pull out the CERT_CONTEXT* and CERT_INFO* pointers without
+                // bringing in all the SafeCertContextHandle machinery, and embed pointers to them in the memory block we return. Yes, this code is scary.
+                using (SafeCertContextHandle hCertContext = recipient.Certificate.CreateCertContextHandle())
+                {
+                    CERT_CONTEXT* pCertContext = hCertContext.DangerousGetCertContext();
+                    CERT_INFO* pCertInfo = pCertContext->pCertInfo;
+
+                    CMSG_KEY_AGREE_RECIPIENT_ENCODE_INFO* pEncodeInfo = (CMSG_KEY_AGREE_RECIPIENT_ENCODE_INFO*)(hb.Alloc(sizeof(CMSG_KEY_AGREE_RECIPIENT_ENCODE_INFO)));
+
+                    pEncodeInfo->cbSize = sizeof(CMSG_KEY_AGREE_RECIPIENT_ENCODE_INFO);
+
+                    pEncodeInfo->KeyEncryptionAlgorithm.pszObjId = hb.AllocAsciiString(Oids.Esdh);
+                    pEncodeInfo->KeyEncryptionAlgorithm.Parameters.cbData = 0;
+                    pEncodeInfo->KeyEncryptionAlgorithm.Parameters.pbData = IntPtr.Zero;
+
+                    pEncodeInfo->pvKeyEncryptionAuxInfo = null;
+
+                    string oidValue;
+                    AlgId algId = contentEncryptionAlgorithm.Oid.Value.ToAlgId();
+                    if (algId == AlgId.CALG_RC2)
+                        oidValue = Oids.CmsRc2Wrap;
+                    else
```

Comment:
I know that this is what desktop does; but this is kinda frightening.  Is there anything that guards that only these two are valid?